client/common/src/types.rs

```diff
@@ -1,4 +1,153 @@
+            return full_chunks + 1;
+        }
+        full_chunks
+    }
+
+    pub fn chunk_ids(&self) -> impl Iterator<Item = ChunkId> {
+        0..self.chunk_count()
+    }
+
+    /// Compute the hash of the SCALE encoded metadata using [`Hasher`].
+    pub fn key<H: Hasher>(&self) -> H::Out {
+        H::hash(&self.encode())
+    }
+
+    /// Decode metadata from the SCALE encoded metadata bytes.
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, codec::Error> {
+        decode_from_bytes(bytes.into())
+    }
+}
+
+/// Typed u64 representing the index of a file [`Chunk`]. Indexed from 0.
+pub type ChunkId = u64;
+
+// TODO: this is currently a placeholder in order to define Storage interface.
+/// Typed chunk of a file. This is what is stored in the leaf of the stored Merkle tree.
+pub type Chunk = Vec<u8>;
+
+/// Leaf in the Forest or File trie.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Leaf<K, D: Debug> {
+    pub key: K,
+    pub data: D,
+}
+
+impl<K, D: Debug> Leaf<K, D> {
+    pub fn new(key: K, data: D) -> Self {
+        Self { key, data }
+    }
+}
+
+/// Proving either the exact key or the neighbour keys of the challenged key.
+pub enum Proven<K, D: Debug> {
+    ExactKey(Leaf<K, D>),
+    NeighbourKeys((Option<Leaf<K, D>>, Option<Leaf<K, D>>)),
+}
+
+impl<K, D: Debug> Proven<K, D> {
+    pub fn new_exact_key(key: K, data: D) -> Self {
+        Proven::ExactKey(Leaf { key, data })
+    }
+
+    pub fn new_neighbour_keys(
+        left: Option<Leaf<K, D>>,
+        right: Option<Leaf<K, D>>,
+    ) -> Result<Self, &'static str> {
+        match (left, right) {
+            (None, None) => Err("Both left and right leaves cannot be None"),
+            (left, right) => Ok(Proven::NeighbourKeys((left, right))),
+        }
+    }
+}
+
+/// Proof of file key(s) in the forest trie.
+pub struct ForestProof<T: TrieLayout> {
+    /// The file key that was proven.
+    pub proven: Vec<Proven<HasherOutT<T>, Metadata>>,
+    /// The compact proof.
+    pub proof: CompactProof,
+    /// The root hash of the trie.
+    pub root: HasherOutT<T>,
+}
+
+/// Storage proof in compact form.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct SerializableCompactProof {
+    pub encoded_nodes: Vec<Vec<u8>>,
+}
+
+impl From<CompactProof> for SerializableCompactProof {
+    fn from(proof: CompactProof) -> Self {
+        Self {
+            encoded_nodes: proof.encoded_nodes,
+        }
+    }
+}
+
+impl Into<CompactProof> for SerializableCompactProof {
+    fn into(self) -> CompactProof {
+        CompactProof {
+            encoded_nodes: self.encoded_nodes,
+        }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct FileProof {
+    /// The file chunk (and id) that was proven.
+    pub proven: Leaf<ChunkId, Chunk>,
+    /// The compact proof.
+    pub proof: SerializableCompactProof,
+    /// The root hash of the trie, also known as the fingerprint of the file.
+    pub root: H256,
```

Comment:
`H256` would be changed to our new type `Hash` as well, but this one can be changed in another PR since it's out of scope.

Reply:
Same as with `ForestProof`, also agree it's out of scope.